File: prototype/node/src/app/core/moderation/listing-removal.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { activeRemoval, canLiftRemoval, type ListingRemoval } from './listing-removal.ts'

function removal(overrides: Partial<ListingRemoval> = {}): ListingRemoval {
  return { kind: 'temporary', reason: 'reported artwork', liftedAt: null, ...overrides }
}

test('no removals means no active removal', () => {
  assert.equal(activeRemoval([]), null)
})

test('an unlifted removal is active', () => {
  const active = removal()
  assert.equal(activeRemoval([active]), active)
})

test('a lifted removal is not active', () => {
  assert.equal(activeRemoval([removal({ liftedAt: new Date('2026-01-01') })]), null)
})

test('the first unlifted removal among several is active', () => {
  const lifted = removal({ liftedAt: new Date('2026-01-01') })
  const active = removal({ reason: 'second report' })
  assert.equal(activeRemoval([lifted, active]), active)
})

test('a temporary removal can be lifted', () => {
  assert.equal(canLiftRemoval('temporary'), true)
})

test('a permanent removal cannot be lifted', () => {
  assert.equal(canLiftRemoval('permanent'), false)
})

File: prototype/rails/src/test/models/listing_removal_test.rb
require "test_helper"

class ListingRemovalTest < ActiveSupport::TestCase
  test "a temporary removal can be lifted" do
    assert_predicate build(kind: :temporary), :liftable?
  end

  test "a permanent removal cannot be lifted" do
    refute_predicate build(kind: :permanent), :liftable?
  end

  test "an unlifted removal is active" do
    assert_predicate build, :active?
  end

  test "a lifted removal is not active" do
    refute_predicate build(lifted_at: Time.current), :active?
  end

  test "active narrows to the removals nobody has lifted" do
    listing = create_listing
    admin = create_admin
    lifted = ListingRemoval.create!(listing: listing, admin: admin, kind: :temporary, reason: "First", lifted_at: Time.current)
    standing = ListingRemoval.create!(listing: listing, admin: admin, kind: :permanent, reason: "Second")

    assert_equal [ standing ], ListingRemoval.active.to_a
    assert_not_includes ListingRemoval.active, lifted
  end

  test "a reason is required" do
    refute_predicate build(reason: ""), :valid?
  end

  private

  def build(**overrides)
    ListingRemoval.new({
      listing: create_listing, admin: create_admin, kind: :temporary, reason: "Reported as counterfeit"
    }.merge(overrides))
  end
end
